✨ Slack integration

enandr opened the pull request:
I have added Slack Integration to update your status on Slack automatically. You can toggle on or off the Slack integration. You can enter your Slack User Token (which you get from creating and installing a custom Slack app to your workspace) and choose the amount of time until your status will be cleared after you pause the music.

enandr committed:
Add Slack integration for updating status

Introduced Slack integration to automatically update the user's Slack status based on the current playing track. Settings for enabling/disabling Slack and configuring the user token and status clear time have been added to the settings UI.

enandr committed:
removed uneded input emit

enandr committed:
added a check to ensure if you have no token, no request is sent

Alipoodle commented on src/main/index.ts:
```diff
@@ -1341,6 +1358,9 @@ app.on("ready", async () => {
         if ("settings-discord-rich-presence" in v1Config) {
           store.set("integrations.discordPresenceEnabled", v1Config["settings-discord-rich-presence"]);
         }
+        if ("settings-slack-status" in v1Config) {
+          store.set("integrations.slackEnabled", v1Config["slack-status"]);
+        }
```
This was not included as a feature in v1, can I ask why this was added?

enandr replied:
I am still learning your codebase. I saw the others using this so I did as well. You are right that Slack is not a v1 feature. I couldn't though find anything for v2Config. Does it currently exist?

Alipoodle replied:
This is a set of Migration code from V1 -> V2, and this is a new feature.

This was not a feature which was in the v1 of the app, so it is not required.

Alipoodle commented on src/main/integrations/slack/index.ts:
```diff
@@ -0,0 +1,81 @@
+
+  private async playerStateChanged(state: PlayerState) {
+    const { secondsTillClearStatus } = await this.store.get("slack");
+    if (state.trackState === 1 && state.videoDetails) {
```
Please can you use the `VideoState` enum instead of the Hard Coded number.

enandr replied:
This has been updated in my latest push.

Alipoodle commented on src/main/integrations/slack/index.ts:
```diff
@@ -0,0 +1,81 @@
+  }
+  private updateSlackStatus(clear: boolean, data: VideoDetails) {
+    const message = `Listening To: ${data.title} | ${data.author}`;
+    let postData: { status_text: string; status_emoji: string; status_expiration: number } = {
```
Feel like there's a cleaning way of doing this Type Checking either with a a new `type.ts` file which includes what the data format is for posting data / recieved Data ETC.

enandr replied:
I have created a new type file and put that type there. that one there was supposed to be temporary that I just forgot to move out. its in my latest push

Alipoodle commented on src/shared/store/schema.ts:
```diff
@@ -52,6 +53,10 @@ export type StoreSchema = {
     scrobblePercent: number;
   };
+  slack: {
+    slackUserToken: string;
```
This is a wider thing than just this file 🤔 
But as this is a "Secret" that this should be Encrypted and Decrypted in the integration.

There would probably be some bits in regards to preventing this being overwritten if left blank in settings 🤷

enandr replied:
It would be better as the user token is a secret and should be encrypted. i will update that in my next push.

What do you mean by "This is a wider thing than just this file"

Alipoodle replied:
You need to grab the encrypted value in your `slack/index.ts` and decrypt it.
And Currently we do not have a proper settings input field which supports for `TextEncrypted` and handles the initial Encryption on the Settings page.

enandr replied:
the encryption is done. i found the way to encrypt it and decrypt it also on the settings page so you can view it

Alipoodle commented on src/renderer/windows/settings/Settings.vue:
```diff
@@ -426,6 +436,29 @@ window.ytmd.handleUpdateDownloaded(() => {
+            type="range"
+            name="Clear Status Seconds"
+            description="Time until the status is automatically cleared"
+            min="5"
+            max="120"
+            step="5"
```
This standard or should we allow a more open range
For Discord's rich presneces we have it hard coded for 60 seconds when paused to hide it? Is it not worth while in doing the same for this?

enandr replied:
I was thinking about hard coding but I decided against it. I have 60 as the default. As Slack is primarily a business platform, not a gaming one like discord, it would be nice to have a range of options.

Alipoodle commented on src/renderer/windows/settings/Settings.vue:
```diff
@@ -426,6 +436,29 @@ window.ytmd.handleUpdateDownloaded(() => {
             @change="settingsChanged"
           />
+          <YTMDSetting v-model="slackEnabled" type="checkbox" name="Slack" @change="settingsChanged" />
+          <i v-if="slackEnabled">You must create a Slack app and generate a user token to use slack integration</i>
```
This feels like it is not needed, but I haven't built or seen the in-app thing to understand how this looks.

A user shouldn't have to be told something like this for a setting. If they are enabling a setting I would hope they would at least have some understanding of what it does, especially considering **Slack** is a bit more of a common thing~

enandr replied:
After second thought, you are right, its not needed. It has been removed in my latest push

Alipoodle commented on src/main/integrations/slack/index.ts:
```diff
@@ -0,0 +1,81 @@
+        status_expiration: newUnixTimestamp
+      };
+    }
+    if (!this.slackUserToken) return;
```
Could this be called earlier to prevent processing of the function earlier?

enandr replied:
You are right. the changes have been reflected in my latest push

Alipoodle commented on src/main/integrations/slack/index.ts:
```diff
@@ -0,0 +1,81 @@
+      status_expiration: 0
+    };
+    if (!clear) {
+      const currentUnixTimestamp = Math.floor(Date.now() / 1000);
```
`currentUnixTimestamp` and `message` feel like they're not really needed as variables if they are only storing stuff for a single usage within the same function?

I would personally just put them directly inside of the `postData`.

enandr replied:
the timestamp can be moved from 2 variable to 1 and message can be removed and used without a variable. this has been reflected in my latest push

enandr replied:
The way slack works though is that you must send empty data to clear the status. any data inside will create a status. I have instead created 2 separate post data variables, one with blank data that clears, one with real data that creates a status

Alipoodle commented on src/main/integrations/slack/index.ts:
```diff
@@ -0,0 +1,81 @@
+    const message = `Listening To: ${data.title} | ${data.author}`;
+    let postData: { status_text: string; status_emoji: string; status_expiration: number } = {
+      status_text: "",
+      status_emoji: "",
```
*Idea*
🎵 Emoji?

enandr replied:
Not a bad idea. I could have a box where you can enter the emoji you wish to use or offer a few options. Right now I use 🎶 hard coded

Alipoodle reviewed (changes requested):
😋 A few changes from a quick skim of the code.

enandr committed:
Changes for review for request 2, 3, 6, 7, 8

enandr committed:
removed slack from v1config

Alipoodle commented on src/main/@types/slack.d.ts:
```diff
@@ -0,0 +1,2 @@
+type PostType = { status_text: string; status_emoji: string; status_expiration: number };
```
1) Can this instead be `export type ...` to achieve both the type Creation + sharing it.
2) is it possible to have some new lines in this, just for readability 😄 
3) Can this be renamed to **SlackPostData** or something.
   1) Prefix it with the Integration name (to not confuse it in Intelisense)
   2) `Type` shouldn't be needed to be included as that is what it's defined as.
   3) PostData is actually what this is being used for.

Alipoodle commented on src/main/integrations/slack/index.ts:
```diff
@@ -0,0 +1,78 @@
+  private async playerStateChanged(state: PlayerState) {
+    if (!this.enabled) return;
+    const { secondsTillClearStatus } = await this.store.get("slack");
+    if (state.trackState === VideoState.Playing && state.videoDetails) {
+      if (!state.adPlaying) {
```
Any reason to not combine these 2 statements or do some early exits instead.
Feels like it would clean up from having a spaghetti of If statements a bit more. ❤

```js
if (!this.enabled ||
    !state.adPlaying ||
    !state.videoDetails ||
    state.trackState !== VideoState.Playing
 ) return;

enandr replied:
i like what you sent, but i did something similar

Alipoodle commented on src/main/integrations/slack/index.ts:
```diff
@@ -0,0 +1,78 @@
+  }
+  private updateSlackStatus(clear: boolean, data: VideoDetails) {
+    const currentUnixTimestamp = Math.floor(Date.now() / 1000);
+    const clearPostData: PostType = {
+      status_text: "",
+      status_emoji: "",
+      status_expiration: 0
+    };
+    const listeningToPostData: PostType = {
+      status_text: `Listening To: ${data.title} | ${data.author}`,
+      status_emoji: ":notes:",
+      status_expiration: currentUnixTimestamp + 15 * 60
+    };
```
Can this become 1 variable instead?
I would play with the formatting a bit to possibly make it a bit more human readable, but something like this is what I am thinking about.
```js
const postData: PostType = (
   clear
   ? {
      status_text: "",
      status_emoji: "",
      status_expiration: 0
    }
    : {
      status_text: `Listening To: ${data.title} | ${data.author}`,
      status_emoji: ":notes:",
      status_expiration: currentUnixTimestamp + 15 * 60
    }
);

enandr replied:
its now instead of this, each keys value is set based on clear or not

Alipoodle reviewed (commented):
2 Comments from the changes you've just made.
and 1 which I forgot to see.

Otherwise there's the setting thing from before, which is a bit more of a task overall. 😅

enandr committed:
Refactor Slack integration to use encrypted user tokens

Updated the Slack integration to use encrypted user tokens and improved the `playerStateChanged` method by combining condition checks. Replaced `PostType` with `SlackPostData` for data consistency. Adjusted settings to handle the encryption and decryption seamlessly.

enandr commented:
@Alipoodle I would love for this to be a part of YTMD and I can then call myself a contributor and do more. I belive all of your review requests are completed
